synapse/storage/state.py

```diff
@@ -532,6 +534,40 @@ def approx_difference(self, other: "StateFilter") -> "StateFilter":
             new_all, new_excludes, new_wildcards, new_concrete_keys
         )
 
+    def must_await_full_state(self, is_mine_id: Callable[[str], bool]) -> bool:
+        """Check if we need to wait for full state to complete to calculate this state
+
+        If we have a state filter which is completely satisfied even with partial
+        state, then we don't need to await_full_state before we can return it.
+
+        Args:
+            is_mine_id: a callable which confirms if a given state_key matches a mxid
+               of a local user
+        """
+
+        # XXX: can we be certain that the state at an event never changes (only gets
+        #            enlarged)?
```

Comment:
kinda both, and yeah, sorry it's not very clear.

The premise here is that if we have a caller which is looking for a given piece of non-membership state (say `m.room.topic`), then it should be safe to return it immediately rather than wait for the state to re-sync.

However, that's not actually a given: it's entirely possible that when we come to recalculate the state, we'll discover that (for example) the `m.room.topic` event was "sent" by someone who wasn't in the room - which means that we previously returned an incorrect result.

It's entirely unclear to me how much of a problem that is going to be in practice.

I'll try to clarify.

Reply:
Thanks, it makes sense now. I somehow missed the effect of the `include_others` option.